### PR TITLE
[SPLIT][SMALL] Fix cost to work more robustly for primal_dual optimization

### PR DESCRIPTION
**Pull request by chaithyagr**
This PR splits up PR #74 . 

We need this merged after which I will rebase #74

### CODE REVIEW

**Comment by chaithyagr** [comment]
The CircleCI is failing as Modopt release has not yet been done..

**Comment by chaithyagr** [comment]
With new Modopt, we are now passing!!

File: mri/optimizers/utils/cost.py [hunk 1]
```diff
@@ -133,5 +142,8 @@ def _calc_cost(self, x_new, *args, **kwargs):
         cost: float
             the cost function defined by the operators (gradient + prox_op).
         """
-        cost = self.gradient_op.cost(x_new) + self.prox_op.cost(x_new)
+        if self.optimizer_type == 'forward_backward':
+            cost = self.gradient_op.cost(x_new) + self.prox_op.cost(x_new)
+        else:
+            cost = self.gradient_op.cost(x_new) + self.prox_op.cost(args[0])
```
Comment (zaccharieramzi) [inline mri/optimizers/utils/cost.py]: are we sure that `args[0]` is always going to be the linear op applied to the optimisation variable ?

in any case it's not very clear hear and deserves at least a comment, but I would be happy if there was a better way.

File: mri/reconstructors/base.py [hunk 2]
```diff
@@ -138,8 +137,16 @@ def reconstruct(self, kspace_data, optimization_alg='pogm',
...
+            optimizer_type = 'primal_dual'
         else:
             kwargs["prox_op"] = self.prox_op
+            optimizer_type = 'forward_backward'
+        self.cost_op = GenericCost(
+            gradient_op=self.gradient_op,
+            prox_op=self.prox_op,
+            verbose=self.verbose >= 20,
+            optimizer_type=optimizer_type
+        )
```
Comment (zaccharieramzi) [inline mri/reconstructors/base.py]: you are not using `cost_op_kwargs`

File: mri/optimizers/utils/cost.py [hunk 3]
```diff
@@ -22,7 +22,7 @@ class DualGapCost(costObj):
     """ Define the dual-gap cost function.
     """
     def __init__(self, linear_op, initial_cost=1e6, tolerance=1e-4,
-                 cost_interval=1, test_range=4, verbose=False,
+                 cost_interval=None, test_range=4, verbose=False,
```
Comment (zaccharieramzi) [inline mri/optimizers/utils/cost.py]: I feel like this is unrelated

Comment (chaithyagr) [reply mri/optimizers/utils/cost.py]: This is just to speed up reconstruction as currently, we calculate the cost at every iteration.

It is related to this PR, only in the sense that we are changing cost related stuff here. 
It is just a single line change.

Comment (zaccharieramzi) [reply mri/optimizers/utils/cost.py]: Makes sense.

File: mri/optimizers/utils/cost.py (hunk 1 above)
Comment (chaithyagr) [reply mri/optimizers/utils/cost.py]: I agree that this is bad and non-scalable code, but for now I could not find a better way to do it. 
I have added a TODO to generalize the code here, also added a comment explaining the code better

Comment (zaccharieramzi) [reply mri/optimizers/utils/cost.py]: ok sure

File: mri/reconstructors/base.py [hunk 4]
```diff
@@ -102,14 +102,10 @@ def initialize_gradient_op(self, **extra_args):
...
         )
-        self.cost_op = GenericCost(
-            gradient_op=self.gradient_op,
-            prox_op=self.prox_op,
-            verbose=self.verbose >= 20,
-        )
 
     def reconstruct(self, kspace_data, optimization_alg='pogm',
-                    x_init=None, num_iterations=100, **kwargs):
+                    x_init=None, num_iterations=100, cost_op_kwargs={},
```
Comment (zaccharieramzi) [inline mri/reconstructors/base.py]: generally we should avoid mutable as defaults, so you could have `None` here and test `if cost_op_kwargs is None: cost_op_kwargs = {}`
